From e7a741e6e471495017d8f6bbf8d73dedf2bf93c9 Mon Sep 17 00:00:00 2001
From: Stephan Bergmann <sbergman@redhat.com>
Date: Thu, 6 Jun 2013 10:10:21 +0200
Subject: [PATCH] rhbz#908819: Don't call code in UNO object ctor that throws
 UNO exceptions

...with Context set to this, that leads to refcounting bugs.  This is a backport
of the relevant parts of 4e42ce3271154904b7478b9ed5e6e6856b9235c2 "Don't call
code in UNO object ctor that throws UNO exceptions."

Change-Id: Ie6fdf92fdf537b752cd05ed69c3b59f2c7795e28
---
 cppuhelper/source/defaultbootstrap.cxx | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/cppuhelper/source/defaultbootstrap.cxx b/cppuhelper/source/defaultbootstrap.cxx
index 7fd1e4a..e027564 100644
--- a/cppuhelper/source/defaultbootstrap.cxx
+++ b/cppuhelper/source/defaultbootstrap.cxx
@@ -590,13 +590,13 @@ class ServiceManager:
     private osl::Mutex, public ServiceManagerBase, private boost::noncopyable
 {
 public:
-    explicit ServiceManager(rtl::OUString const & rdbUris):
-        ServiceManagerBase(*static_cast< osl::Mutex * >(this))
-    { readRdbs(rdbUris); }
+    ServiceManager(): ServiceManagerBase(*static_cast< osl::Mutex * >(this)) {}
 
     using ServiceManagerBase::acquire;
     using ServiceManagerBase::release;
 
+    void init(rtl::OUString const & rdbUris) { readRdbs(rdbUris); }
+
     void setContext(
         css::uno::Reference< css::uno::XComponentContext > const & context)
     {
@@ -1996,7 +1996,8 @@ css::uno::Reference< css::uno::XComponentContext > bootstrapComponentContext(
     css::uno::Reference< css::registry::XSimpleRegistry > const & typeRegistry,
     rtl::OUString const & serviceUris, rtl::Bootstrap const & bootstrap)
 {
-    rtl::Reference< ServiceManager > smgr(new ServiceManager(serviceUris));
+    rtl::Reference< ServiceManager > smgr(new ServiceManager);
+    smgr->init(serviceUris);
     cppu::ContextEntry_Init entry;
     std::vector< cppu::ContextEntry_Init > context_values;
     context_values.push_back(
-- 
1.8.1.4

